File: nodejs/test/e2e/client.test.ts
import { ChildProcess } from "child_process";
import { describe, expect, it, onTestFinished } from "vitest";
import { CopilotClient } from "../../src/index.js";

function onTestFinishedForceStop(client: CopilotClient) {
    onTestFinished(async () => {
        try {
            await client.forceStop();
        } catch {
            // Ignore cleanup errors - process may already be stopped
        }
    });
}

describe("Client", () => {
    it("should start and connect to server using stdio", async () => {
        const client = new CopilotClient({ useStdio: true });
        onTestFinishedForceStop(client);

        await client.start();
        expect(client.getState()).toBe("connected");

        const pong = await client.ping("test message");
        expect(pong.message).toBe("pong: test message");
        expect(pong.timestamp).toBeGreaterThanOrEqual(0);

        expect(await client.stop()).toHaveLength(0); // No errors on stop
        expect(client.getState()).toBe("disconnected");
    });

    it("should start and connect to server using tcp", async () => {
        const client = new CopilotClient({ useStdio: false });
        onTestFinishedForceStop(client);

        await client.start();
        expect(client.getState()).toBe("connected");

        const pong = await client.ping("test message");
        expect(pong.message).toBe("pong: test message");
        expect(pong.timestamp).toBeGreaterThanOrEqual(0);

        expect(await client.stop()).toHaveLength(0); // No errors on stop
        expect(client.getState()).toBe("disconnected");
    });

    it.skipIf(process.platform === "darwin")("should return errors on failed cleanup", async () => {
        // Use TCP mode to avoid stdin stream destruction issues
        // Without this, on macOS there are intermittent test failures
        // saying "Cannot call write after a stream was destroyed"
        // because the JSON-RPC logic is still trying to write to stdin after
        // the process has exited.
        const client = new CopilotClient({ useStdio: false });

        await client.createSession();

        // Kill the server process to force cleanup to fail
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const cliProcess = (client as any).cliProcess as ChildProcess;
        expect(cliProcess).toBeDefined();
        cliProcess.kill("SIGKILL");
        await new Promise((resolve) => setTimeout(resolve, 100));

        const errors = await client.stop();
        expect(errors.length).toBeGreaterThan(0);
        expect(errors[0].message).toContain("Failed to destroy session");
    });

    it("should forceStop without cleanup", async () => {
        const client = new CopilotClient({});
        onTestFinishedForceStop(client);

        await client.createSession();
        await client.forceStop();
        expect(client.getState()).toBe("disconnected");
    });

    it("should get status with version and protocol info", async () => {
        const client = new CopilotClient({ useStdio: true });
        onTestFinishedForceStop(client);

        await client.start();

        const status = await client.getStatus();
        expect(status.version).toBeDefined();
        expect(typeof status.version).toBe("string");
        expect(status.protocolVersion).toBeDefined();
        expect(typeof status.protocolVersion).toBe("number");
        expect(status.protocolVersion).toBeGreaterThanOrEqual(1);

        await client.stop();
    });

    it("should get auth status", async () => {
        const client = new CopilotClient({ useStdio: true });
        onTestFinishedForceStop(client);

        await client.start();

        const authStatus = await client.getAuthStatus();
        expect(typeof authStatus.isAuthenticated).toBe("boolean");
        if (authStatus.isAuthenticated) {
            expect(authStatus.authType).toBeDefined();
            expect(authStatus.statusMessage).toBeDefined();
        }

        await client.stop();
    });

    it("should list models when authenticated", async () => {
        const client = new CopilotClient({ useStdio: true });
        onTestFinishedForceStop(client);

        await client.start();

        const authStatus = await client.getAuthStatus();
        if (!authStatus.isAuthenticated) {
            // Skip if not authenticated - models.list requires auth
            await client.stop();
            return;
        }

        const models = await client.listModels();
        expect(Array.isArray(models)).toBe(true);
        if (models.length > 0) {
            const model = models[0];
            expect(model.id).toBeDefined();
            expect(model.name).toBeDefined();
            expect(model.capabilities).toBeDefined();
            expect(model.capabilities.supports).toBeDefined();
            expect(model.capabilities.limits).toBeDefined();
        }

        await client.stop();
    });
});

File: go/internal/e2e/client_test.go
package e2e

import (
	"testing"
	"time"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

func TestClient(t *testing.T) {
	cliPath := testharness.CLIPath()
	if cliPath == "" {
		t.Fatal("CLI not found. Run 'npm install' in the nodejs directory first.")
	}

	t.Run("should start and connect to server using stdio", func(t *testing.T) {
		client := copilot.NewClient(&copilot.ClientOptions{
			CLIPath:  cliPath,
			UseStdio: copilot.Bool(true),
		})
		t.Cleanup(func() { client.ForceStop() })

		if err := client.Start(t.Context()); err != nil {
			t.Fatalf("Failed to start client: %v", err)
		}

		if client.State() != copilot.StateConnected {
			t.Errorf("Expected state to be 'connected', got %q", client.State())
		}

		pong, err := client.Ping(t.Context(), "test message")
		if err != nil {
			t.Fatalf("Failed to ping: %v", err)
		}

		if pong.Message != "pong: test message" {
			t.Errorf("Expected pong.message to be 'pong: test message', got %q", pong.Message)
		}

		if pong.Timestamp < 0 {
			t.Errorf("Expected pong.timestamp >= 0, got %d", pong.Timestamp)
		}

		if err := client.Stop(); err != nil {
			t.Errorf("Expected no errors on stop, got %v", err)
		}

		if client.State() != copilot.StateDisconnected {
			t.Errorf("Expected state to be 'disconnected', got %q", client.State())
		}
	})

	t.Run("should start and connect to server using tcp", func(t *testing.T) {
		client := copilot.NewClient(&copilot.ClientOptions{
			CLIPath:  cliPath,
			UseStdio: copilot.Bool(false),
		})
		t.Cleanup(func() { client.ForceStop() })

		if err := client.Start(t.Context()); err != nil {
			t.Fatalf("Failed to start client: %v", err)
		}

		if client.State() != copilot.StateConnected {
			t.Errorf("Expected state to be 'connected', got %q", client.State())
		}

		pong, err := client.Ping(t.Context(), "test message")
		if err != nil {
			t.Fatalf("Failed to ping: %v", err)
		}

		if pong.Message != "pong: test message" {
			t.Errorf("Expected pong.message to be 'pong: test message', got %q", pong.Message)
		}

		if pong.Timestamp < 0 {
			t.Errorf("Expected pong.timestamp >= 0, got %d", pong.Timestamp)
		}

		if err := client.Stop(); err != nil {
			t.Errorf("Expected no errors on stop, got %v", err)
		}

		if client.State() != copilot.StateDisconnected {
			t.Errorf("Expected state to be 'disconnected', got %q", client.State())
		}
	})

	t.Run("should return errors on failed cleanup", func(t *testing.T) {
		client := copilot.NewClient(&copilot.ClientOptions{
			CLIPath: cliPath,
		})
		t.Cleanup(func() { client.ForceStop() })

		_, err := client.CreateSession(t.Context(), nil)
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		// Kill the server process to force cleanup to fail
		client.ForceStop()
		time.Sleep(100 * time.Millisecond)

		if err := client.Stop(); err != nil {
			t.Logf("Got expected errors: %v", err)
		}

		if client.State() != copilot.StateDisconnected {
			t.Errorf("Expected state to be 'disconnected', got %q", client.State())
		}
	})

	t.Run("should forceStop without cleanup", func(t *testing.T) {
		client := copilot.NewClient(&copilot.ClientOptions{
			CLIPath: cliPath,
		})
		t.Cleanup(func() { client.ForceStop() })

		_, err := client.CreateSession(t.Context(), nil)
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		client.ForceStop()

		if client.State() != copilot.StateDisconnected {
			t.Errorf("Expected state to be 'disconnected', got %q", client.State())
		}
	})

	t.Run("should get status with version and protocol info", func(t *testing.T) {
		client := copilot.NewClient(&copilot.ClientOptions{
			CLIPath:  cliPath,
			UseStdio: copilot.Bool(true),
		})
		t.Cleanup(func() { client.ForceStop() })

		if err := client.Start(t.Context()); err != nil {
			t.Fatalf("Failed to start client: %v", err)
		}

		status, err := client.GetStatus(t.Context())
		if err != nil {
			t.Fatalf("Failed to get status: %v", err)
		}

		if status.Version == "" {
			t.Error("Expected status.Version to be non-empty")
		}

		if status.ProtocolVersion < 1 {
			t.Errorf("Expected status.ProtocolVersion >= 1, got %d", status.ProtocolVersion)
		}

		client.Stop()
	})

	t.Run("should get auth status", func(t *testing.T) {
		client := copilot.NewClient(&copilot.ClientOptions{
			CLIPath:  cliPath,
			UseStdio: copilot.Bool(true),
		})
		t.Cleanup(func() { client.ForceStop() })

		if err := client.Start(t.Context()); err != nil {
			t.Fatalf("Failed to start client: %v", err)
		}

		authStatus, err := client.GetAuthStatus(t.Context())
		if err != nil {
			t.Fatalf("Failed to get auth status: %v", err)
		}

		// isAuthenticated is a bool, just verify we got a response
		if authStatus.IsAuthenticated {
			if authStatus.AuthType == nil {
				t.Error("Expected authType to be set when authenticated")
			}
			if authStatus.StatusMessage == nil {
				t.Error("Expected statusMessage to be set when authenticated")
			}
		}

		client.Stop()
	})

	t.Run("should list models when authenticated", func(t *testing.T) {
		client := copilot.NewClient(&copilot.ClientOptions{
			CLIPath:  cliPath,
			UseStdio: copilot.Bool(true),
		})
		t.Cleanup(func() { client.ForceStop() })

		if err := client.Start(t.Context()); err != nil {
			t.Fatalf("Failed to start client: %v", err)
		}

		authStatus, err := client.GetAuthStatus(t.Context())
		if err != nil {
			t.Fatalf("Failed to get auth status: %v", err)
		}

		if !authStatus.IsAuthenticated {
			// Skip if not authenticated - models.list requires auth
			client.Stop()
			return
		}

		models, err := client.ListModels(t.Context())
		if err != nil {
			t.Fatalf("Failed to list models: %v", err)
		}

		if len(models) > 0 {
			model := models[0]
			if model.ID == "" {
				t.Error("Expected model.ID to be non-empty")
			}
			if model.Name == "" {
				t.Error("Expected model.Name to be non-empty")
			}
		}

		client.Stop()
	})

	t.Run("should report error when CLI fails to start", func(t *testing.T) {
		client := copilot.NewClient(&copilot.ClientOptions{
			CLIPath:  cliPath,
			CLIArgs:  []string{"--nonexistent-flag-for-testing"},
			UseStdio: copilot.Bool(true),
		})
		t.Cleanup(func() { client.ForceStop() })

		err := client.Start(t.Context())
		if err == nil {
			t.Fatal("Expected Start to fail with invalid CLI args")
		}

		// Verify subsequent calls also fail (don't hang)
		session, err := client.CreateSession(t.Context(), nil)
		if err == nil {
			_, err = session.Send(t.Context(), copilot.MessageOptions{Prompt: "test"})
		}
		if err == nil {
			t.Fatal("Expected CreateSession/Send to fail after CLI exit")
		}
	})
}
